alf/utils/data_buffer.py

```diff
@@ -287,46 +287,77 @@ def dequeue(self, env_ids=None, blocking=False):
+        """Return earliest ``n`` steps and mark them removed in the buffer.
 
         Args:
             env_ids (Tensor): If None, ``batch_size`` must be num_environments.
                 If not None, dequeue from these environments. We assume there
                 is no duplicate ids in ``env_id``. ``result[i]`` will be from
                 environment env_ids[i].
+            n (int): Number of steps to dequeue.
         Returns:
-            nested Tensors of shapes ``[batch_size, ...]``.
+            nested Tensors of shape ``[batch_size, n, ...]``.
         Raises:
             ``AssertionError`` when not enough data is present.
         """
         with alf.device(self._device):
             env_ids = self.check_convert_env_ids(env_ids)
             current_size = self._current_size[env_ids]
             min_size = current_size.min()
-            assert min_size >= 1, (
+            assert min_size >= n, (
                 "Not all environments have enough data. The smallest data "
                 "size is: %s Try storing more data before calling dequeue" %
                 min_size)
-            pos = self._current_pos[env_ids] - current_size
-            pos = pos % self._max_length
-            indices = env_ids * self._max_length + pos  # shape [B]
-
             batch_size = env_ids.shape[0]
-            result = alf.nest.map_structure(
-                lambda buffer: buffer[indices].reshape(
-                    batch_size, 1, *buffer.shape[1:]), self._flattened_buffer)
-
-            self._current_size[env_ids] = current_size - 1
+            pos = self._current_pos[env_ids] - current_size  # mod done later
+            b_indices = env_ids.reshape(batch_size, 1).expand(-1, n)
+            t_range = torch.arange(n).reshape(1, -1)
+            t_indices = (
+                pos.reshape(batch_size, 1) + t_range) % self._max_length
+            result = self.get_data_by_indices(b_indices, t_indices, batch_size,
+                                              n)
+            self._current_size[env_ids] = current_size - n
             # set flags if they exist to unblock potential consumers
             if self._dequeued:
                 self._dequeued.set()
                 self._enqueued.clear()
         return convert_device(result)
 
+    def get_data_by_indices(self, b_indices, t_indices, batch_size, n):
+        """Return data by batch and time indices.
+
+        Args:
+            b_indices (Tensor): Batch indices of shape ``[B, n]``.
+            t_indices (Tensor): Time indices of shape ``[B, n]``.
+            batch_size (int): Number of batches to retrieve.
+                Should equal ``*_indices.shape[0]``.
+            n (int): Number of time steps to retrieve.
+                Should equal ``*_indices.shape[1]``.
+        Returns:
+            nested Tensors of shape ``[batch_size, n, ...]``.
+        """
+        return alf.nest.map_structure(
+            lambda buffer: buffer[(b_indices, t_indices)].reshape(
```

Comment:
perhaps reshape() is unnecessary.

Reply:
and if so, there's not much sense to have this function because it's shot enough to write:
```python
alf.nest.map_structure(lambda b: b[(b_indices, t_indices)], self._buffer)
```